text/0028-generalized-entity-data-source.md

```diff
@@ -0,0 +1,569 @@
+`Value` and based on the Core `Value`. This provides maximum API cleanliness
+but requires the most breaking changes.
+- expose `Value`, but keep `EvalResult` as a type synonym for `Value`. This
+is almost the same as the above option, but keeps the `EvalResult` type name
+around, even though it makes breaking changes to the structure of
+`EvalResult`. This might reduce confusion for users who are migrating, as
+they only have to adapt to the changes to `EvalResult` and not rename
+`EvalResult` to `Value` in their own code; but, it might increase confusion
+for future users, who will see two names for the same thing.
+- not expose `Value` and just use `EvalResult`, even in the new constructors.
+This provides maximum backwards compatibility, like the first option, and
+also avoids the possible confusion generated by the first option, but comes
+at a notable performance cost (due to the inefficient
+`EvalResult`-to-`Value` conversion required under the hood) and might
+generate a different kind of confusion as `EvalResult` is an awkward name in
+the context of these constructors.
+
+This RFC currently proposes we take the third option.
+
+Fourth and finally, we remove `Entities::evaluate()`, as all `Entities` are now
+stored in their evaluated form by default.
+This is not a breaking change because `Entities::evaluate()` has not yet been
+released in any Cedar release.
+
+## Drawbacks
+
+- This RFC represents some breaking changes for users upgrading to a new major
+version of Cedar. All breaking changes come with some cost in terms of user
+experience for existing users. This RFC contends that the benefit outweighs the
+cost in this case.
+
+- `EntityDataSource` represents a new contract we're making with users.  We
+could be locking ourselves in to some implementation choices that currently we'd
+be free to change behind the scenes. Similarly with exposing (even a wrapped
+version of) `Value`.
+
+- Some errors will occur at different times, or more lazily. In particular,
+today, if any entity attribute is invalid, that error will _always_ be caught
+during any call of `is_authorized()`. With this RFC, if an entity is never used
+during an evaluation, its attributes may never be evaluated and thus may not
+cause an error. Changing the behavior of some `is_authorized()` requests from
+"error" to "valid" is not technically a breaking change (otherwise, adding any
+new language syntax would be a breaking change), but this still might surprise
+some users.
+
+- The current proposed definition of `EntityDataSource` assumes that every
+implementation of the trait handles transitive closure properly, and has no way
+to check this assumption. If users make mistakes in this area, they may silently
+get the wrong authorization results.
+(This is also true of implementation errors such as loading the wrong attribute
+values, etc; but errors with TC may feel more insidious somehow.)
```

Comment:
We may not be able to provide a purely drop-in series of tests.  For instance, we're looking for a parent relationship A -> B -> C so that we can test that A -> C also holds.  The `EntityDataSource` API doesn't have any way for the test to discover such an A, B, and C without the user helping (by writing part of the test), nor does it contain mutating APIs that would allow our test harness code to request the appopriate A, B, and C to the user's backing entity store.

Reply:
Hm good point. Should we prescribe a CRUD API? What would be the pros/cons generally of doing that?

Reply:
I'm not sure we should have a CRUD API.  For instance what would this mean for an `EntityDataSource` implementation that's just a read-only cache, or read-only replica, of some other database.  And users wouldn't want our tests to make permanent changes to their database for testing purposes.  Also, the main Cedar evaluator has no reason to use CRUD and it seems weird to add it just to accommodate testing.

Reply:
Good point. Then we have to be crystal clear in the documentation.